src/shared/directives/substitutor/services/placeholder.service.ts

```diff
@@ -18,58 +19,91 @@ ___
       ) { }
 
     resolvePlaceholders(pageFormFields, stringToResolve): string {
-        let startSubstitutionIndex = -1;
-        let fieldIdToSubstitute = '';
-        let isCollecting = false;
-        let originalStringToResolve = stringToResolve;
-        let numberCollectionItemsAsPlaceholder = 1;
-        let colItemIndex = 0;
-        let scanIndex = 0;
-
-        if (stringToResolve && typeof stringToResolve === 'string') {
-            while (numberCollectionItemsAsPlaceholder-- > 0) {
-
-                while (scanIndex < stringToResolve.length) {
-                    if (this.isStartPlaceholderAndNotCollecting(stringToResolve, scanIndex, isCollecting)) {
-                        startSubstitutionIndex = scanIndex;
-                        isCollecting = true;
-                    } else if (isCollecting) {
-                        if (this.isClosingPlaceholder(stringToResolve, scanIndex)) {
-                            if (this.isMatchingPlaceholderPattern(fieldIdToSubstitute)
-                                && this.isFieldIdInFormFields(fieldIdToSubstitute, pageFormFields, colItemIndex)) {
-
-                                    if (fieldIdToSubstitute.split('.').length > 1) {
-                                        let newNumberOfCollectionItemsAsPlaceholder =
-                                            this.getNumberOfCollectionItemsIfAny(pageFormFields, fieldIdToSubstitute);
-                                            numberCollectionItemsAsPlaceholder = this.getNewNumberOfCollectionItemsIfHigher(
-                                                newNumberOfCollectionItemsAsPlaceholder,
-                                                numberCollectionItemsAsPlaceholder)
-                                    }
-                                    stringToResolve = this.substitute(
-                                        pageFormFields, stringToResolve, startSubstitutionIndex, fieldIdToSubstitute, colItemIndex);
-
-                                    scanIndex = this.resetScanIndexAfterSubstitution(
-                                        startSubstitutionIndex, pageFormFields, fieldIdToSubstitute, colItemIndex);
+
+        while (this.hasUnresolvedPlaceholder(stringToResolve)) {
```

Comment:
This code should further be refactored to have no more than 2 or 3 levels of conditional control in a function.
Appreciate the original code was already problematic in this aspect, and if refactoring now is too much beyond the purpose of the particular change, let's then define a new refactoring ticket just so that we don't miss this tech debt off our future agenda.

Reply:
Ticket created: https://tools.hmcts.net/jira/browse/RDM-6248

Reply:
Closed the above as was done as part of this ticket now.